src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -24,65 +24,137 @@
 
 package org.jenkinsci.plugins.workflow.actions;
 
+import com.google.common.primitives.Primitives;
+import org.apache.commons.collections.CollectionUtils;
 import org.jenkinsci.plugins.workflow.graph.FlowNode;
 import org.jenkinsci.plugins.workflow.graph.StepNode;
 import org.jenkinsci.plugins.workflow.steps.Step;
 import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
 
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 /**
  * Stores some or all of the arguments used to create and configure the {@link Step} executed by a {@link FlowNode}.
  * This allows you to inspect information supplied in the pipeline script and otherwise discarded at runtime.
  * Supplied argument values can be hidden and replaced with a {@link NotStoredReason} for security or performance.
+ *
+ * Important note: these APIs do not provide recursive guarantees that returned datastructures are immutable.
  */
 public abstract class ArgumentsAction implements PersistentAction {
 
-    /** Used as a placeholder for arguments not stored for various reasons */
+    /** Used as a placeholder marker for {@link Step} arguments not stored for various reasons. */
     public enum NotStoredReason {
-        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info */
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info. */
         MASKED_VALUE,
 
-        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_STRING_LENGTH} */
+        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_RETAINED_LENGTH} */
         OVERSIZE_VALUE
     }
 
-    /** Largest string we'll persist in the Step, for performance reasons */
-    public static final int MAX_STRING_LENGTH = 1024;
+    /** Largest String, Collection, or array size we'll retain -- provides a rough size limit on any single field.
```

Comment:
For `String` lengths, would be good to clarify that you are referring to character count, not byte count.